stdlib/public/core/StringGuts.swift

```diff
@@ -982,11 +995,24 @@ extension _StringGuts {
     self.append(other._wholeString._guts, range: other._encodedOffsetRange)
   }
 
-  @usableFromInline // @testable
+  @inlinable // @testable
   internal
   mutating func append(_ other: _StringGuts) {
+    // We inline only the _isEmptySingleton check because it can often be
+    // proven or disproven at compile time. A full length check is often
+    // inconclusive.
+    if _isEmptySingleton {
+      self = other
+      return
+    }
+    _appendSlow(other)
+  }
```

Comment:
Currently, `DefaultStringInterpolation.write(_:)` is calling `_appendSlow(_:)` directly to bypass `append(_:)`'s empty-string check, which is false for the majority of segments. This seemed to benchmark well, but it might be over-fitting the benchmarks and/or an unnecessary complication. Or...